drivers/mipi_dbi/mipi_dbi_bflb.c

```diff
@@ -0,0 +1,776 @@
+
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/drivers/mipi_dbi.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/cache.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mipi_dbi_bflb, CONFIG_MIPI_DBI_LOG_LEVEL);
+
+#include <soc.h>
+#include <bflb_soc.h>
+#include <glb_reg.h>
+#include <hbn_reg.h>
+#include <common_defines.h>
+#include <zephyr/dt-bindings/clock/bflb_clock_common.h>
+#include <zephyr/drivers/clock_control/clock_control_bflb_common.h>
+#include <zephyr/dt-bindings/clock/bflb_bl61x_clock.h>
+#include <bouffalolab/common/dbi_reg.h>
+#include <bouffalolab/common/dma_reg.h>
+
+#define DBI_MAX_FREQ		MHZ(80)
+#define DBI_MAX_XCLK_FREQ	MHZ(20)
+#define DBI_MAX_INPUT_FREQ	MHZ(160)
+#define DBI_WAIT_TIMEOUT_MS	500
+
+#define DBI_MODE_QSPI	3U
+#define DBI_MODE_C_3W	2U
+#define DBI_MODE_C_4W	1U
+#define DBI_MODE_B_8B	0U
+
+#define DBI_FIFO_FORMAT_BGR	0U
+#define DBI_FIFO_FORMAT_RGB	1U
+#define DBI_FIFO_FORMAT_RnBGR	4U
+#define DBI_FIFO_FORMAT_BnRGB	5U
+
+#define DBI_MAX_CMD_WRITE_LEN	256
+
+struct mipi_dbi_bflb_config {
+	uint32_t base;
+	const struct pinctrl_dev_config *pincfg;
+	void (*irq_config_func)(const struct device *dev);
+	const struct gpio_dt_spec reset_gpio;
+	const struct gpio_dt_spec dc_gpio;
+	const struct gpio_dt_spec *cs_gpios;
+	size_t num_cs_gpios;
+};
+
+struct mipi_dbi_bflb_dma_data {
+	const struct device *dev;
+	int channel;
+	struct dma_config config;
+	struct dma_block_config block;
+	uint8_t	*buf;
+	size_t	len;
+};
+
+struct mipi_dbi_bflb_data {
+	struct mipi_dbi_config configured;
+	struct k_mutex lock;
+	struct mipi_dbi_bflb_dma_data dma;
+	bool transfer_done;
+};
+
+static inline bool mipi_dbi_has_pin(const struct gpio_dt_spec *spec)
+{
+	return spec->port != NULL;
+}
+
+static inline void mipi_dbi_bflb_cs_control(const struct device *dev,
+					    const struct mipi_dbi_config *config,
+					    bool on, bool force_off)
+{
+	const struct mipi_dbi_bflb_config *cfg = dev->config;
+
+	if (config->config.slave < cfg->num_cs_gpios) {
+		if (mipi_dbi_has_pin(&cfg->cs_gpios[config->config.slave])) {
+			if (on) {
+				gpio_pin_set_dt(&cfg->cs_gpios[config->config.slave], 1);
+				k_busy_wait(config->config.cs.delay);
+			} else {
+				if (!force_off
+				    && (config->config.operation & SPI_HOLD_ON_CS
+					|| IS_ENABLED(CONFIG_MIPI_DBI_BFLB_ASSUME_HOLD_ON_CS))) {
+					return;
+				}
+				k_busy_wait(config->config.cs.delay);
+				gpio_pin_set_dt(&cfg->cs_gpios[config->config.slave], 0);
+			}
+		}
+	}
+}
+
+static bool mipi_dbi_bflb_bus_busy(const struct device *dev)
+{
+	const struct mipi_dbi_bflb_config *config = dev->config;
+	volatile uint32_t tmp;
+
+	tmp = sys_read32(config->base + DBI_CONFIG_OFFSET);
+
+	return (tmp & DBI_STS_DBI_BUS_BUSY) != 0;
+}
+
+static void mipi_dbi_bflb_clear_fifo(const struct device *dev)
+{
+	const struct mipi_dbi_bflb_config *config = dev->config;
+	volatile uint32_t tmp;
```

Comment:
Does this need to be volatile?

Reply:
@danieldegrasse So, any thoughts on this?